velero-plugins/imagestreamtag/restore.go

```diff
@@ -0,0 +1,123 @@
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+// RestorePlugin is a restore item action plugin for Velero
+type RestorePlugin struct {
+	Log logrus.FieldLogger
+}
+
+// AppliesTo returns a velero.ResourceSelector that applies to everything
+func (p *RestorePlugin) AppliesTo() (velero.ResourceSelector, error) {
+	return velero.ResourceSelector{
+		IncludedResources: []string{"imagestreamtags"},
+	}, nil
+}
+
+func (p *RestorePlugin) Execute(input *velero.RestoreItemActionExecuteInput) (*velero.RestoreItemActionExecuteOutput, error) {
+	p.Log.Info("[istag-restore] Hello from ImageStreamTag RestorePlugin!")
+	imageStreamTag := imagev1API.ImageStreamTag{}
+	itemMarshal, _ := json.Marshal(input.Item)
+	json.Unmarshal(itemMarshal, &imageStreamTag)
+	annotations := imageStreamTag.Annotations
+	if annotations == nil {
+		annotations = make(map[string]string)
+	}
+
+	p.Log.Info(fmt.Sprintf("[istag-restore] Restoring imagestreamtag %s", imageStreamTag.Name))
+
+	internalRegistry := annotations[common.RestoreRegistryHostname]
+	if len(internalRegistry) == 0 {
+		return nil, errors.New("restore cluster registry not found for annotation \"openshift.io/restore-registry-hostname\"")
+	}
+	backupInternalRegistry := annotations[common.BackupRegistryHostname]
+	if len(internalRegistry) == 0 {
+		return nil, errors.New("backup cluster registry not found for annotation \"openshift.io/backup-registry-hostname\"")
+	}
+	p.Log.Info(fmt.Sprintf("[istag-restore] backup internal registry: %#v", backupInternalRegistry))
+	dockerImageReference := imageStreamTag.Image.DockerImageReference
+	localImage := common.HasImageRefPrefix(dockerImageReference, backupInternalRegistry)
+	if localImage {
+		p.Log.Info(fmt.Sprintf("[istag-restore] Local image: %v", dockerImageReference))
+	}
+	referenceTag := imageStreamTag.Tag != nil && imageStreamTag.Tag.From != nil
+	var additionalItems []velero.ResourceIdentifier
+	if referenceTag {
+		p.Log.Info(fmt.Sprintf("[istag-restore] Reference tag: %v, tag: %v", imageStreamTag.Tag.From.Kind, imageStreamTag.Tag.From.Name))
+		if imageStreamTag.Tag.From.Kind == "ImageStreamTag" {
+			p.Log.Info("[istag-restore] ImageStreamTag reference")
+			refNamespace := imageStreamTag.Namespace
+			if imageStreamTag.Tag.From.Namespace != "" {
+				refNamespace = imageStreamTag.Tag.From.Namespace
+			}
+			p.Log.Info(fmt.Sprintf("[istag-restore] Looking up reference tag: %s/%s", refNamespace, imageStreamTag.Tag.From.Name))
+			client, err := clients.NewImageClient()
+			_, err = client.ImageStreamTags(refNamespace).Get(imageStreamTag.Tag.From.Name, metav1.GetOptions{})
+			if err == nil {
+				p.Log.Info("[istag-restore] reference tag found in cluster")
+			} else {
+				p.Log.Info("[istag-restore] reference tag not found in cluster, adding to restore")
+				addImageStream := false
+				nameSplit := strings.SplitN(imageStreamTag.Tag.From.Name, ":", 2)
+				if len(nameSplit) == 2 {
```

Comment:
If `len(nameSplit) != 2` do we need to error out?

Reply:
Well, in theory it shouldn't happen. I didn't want to error out the whole thing based on an attempt to pull in dependent items, though. It feels like the wrong place to raise an error. If we need to error out for it here, we need to check it for all imagestreamtags. The only reason I'm checking it here is if I don't have the split, then I have no idea how I'd pull in an ImageStream here.